### ISSUE TITLE
Add context support in server.xml

### ISSUE THREAD

**Issue by hanej** [antoineco/aco-tomcat#19]
We're migrating from WebLogic to Tomcat and our workflow requires the ability to define a context element under the host in the server.xml.  Is there a way to reuse the existing context parameters and apply those to the Host container?  We've tried numerous permutations of context.xml and server.xml and cannot get the application to load properly.

**Comment by hanej** [comment]
Well not reuse the parameters themselves but we could probably reuse the templates.  The indentation may be off a little though.

**Comment by hanej** [comment]
Pull request #21

**Comment by antoineco** [comment]
That's not easy because then to stay consistent we'd have to also offer the possibility to nest contexts anywhere. 

Some more complex modules do it very well, like the ones from Evenup and Puppetlabs. Also if you have a consistent configuration accross your servers you could also distribute a custom server.xml.

**Comment by antoineco** [comment]
Another note: it's not recommended to use contexts inside server.xml, as the official documentation mentions here [Defining a context](https://tomcat.apache.org/tomcat-8.0-doc/config/context.html#Defining_a_context)

However it's super easy to define a "context" defined type with a `path` parameter, which you could create inside `$CATALINA_BASE/conf/[enginename]/[hostname]/`. Quick and elegant way to create contexts for a specific host.

**Comment by hanej** [comment]
I looked at those other projects but they don't have all of the features I need.

We have X number of Tomcat instances sharing the same war file so we put the war file in a common directory.  Tomcat cannot find the war file unless we put the context node inside the server.xml.  I did see that comment about contexts inside the server.xml but if you look further down at docBase, there's a case for doing so.  That is what we need to do.

I remember trying that before and Tomcat didn't see the war file but I'll give it another shot.

**Comment by antoineco** [comment]
That should work as long as the file is named `conf/Catalina/<host_name>/<your_context_path>.xml`, or `conf/Catalina/<host_name>/ROOT.xml` for the `/` context path.

I'm already working on the feature.

**Comment by hanej** [comment]
Awesome.  Thank you.  Our docBase points to a symlink to the correct version of the war.  There are several versions in the same directory to allow for easy downgrades if necessary.  WebLogic has no issue with this but Tomcat does. That's the crux of the situation.

**Comment by antoineco** [comment]
Allright, in this case your context file name should be the same as your symlink base name. When do you think you can test that?

**Comment by antoineco** [comment]
Oh, and just in case you haven't seen it, the same kind of configuration is used within this module to enable the tomcat manager inside instances: [manager.xml.erb](https://github.com/antoineco/aco-tomcat/blob/master/templates/instance/manager.xml.erb)

The file is dropped inside, say, `$CATALINA_BASE/conf/Catalina/myhost.com/manager.xml` (https://github.com/antoineco/aco-tomcat/blob/master/manifests/instance.pp#L1105) and the app is loaded. Should work equally with a war file or directory.

**Comment by hanej** [comment]
Cool.  I'll test it out.  Thanks

**Comment by hanej** [comment]
That worked!  Thank you again.  I don't think I originally had the context file name the same as the symlink.

**Comment by antoineco** [comment]
Great! Let's see if you can achieve the same thing with the new `tomcat::context` type now :)

**Comment by hanej** [comment]
Looks like you're missing some commas :-)

https://github.com/antoineco/aco-tomcat/blob/master/manifests/instance.pp#L874-L885

**Comment by hanej** [comment]
This is a much nicer solution than the one I had.  It makes more sense.

I'm running into two issues though.  One of them I'm still trying to track down the cause but the other one is due to the path parameter in the context.pp manifest.

Our Tomcat instances are all dynamically generated based on a hiera file.  I'm using create_resources to create the application context files but I don't know what the path is until runtime (recursive functions).  Since I can't loop through the hashes to inject the full path for each context file I can only pass in `$CATALINA_BASE/conf/Catalina/localhost` which doesn't work for the concat function because it needs a full path to a file.  Would it make sense to append the `${name}.xml` to the path if the path doesn't end with ".xml"?  That's the only way I could get it to work.

https://github.com/antoineco/aco-tomcat/blob/master/manifests/context.pp#L30
`path   => "${path}/${name}.xml",`

**Comment by hanej** [comment]
I fixed the other issue so now it's just down to the application context file path.  This is the solution I used for context.pp.

```
    # parameters validation
    validate_absolute_path($path)

    $path_real = $path ? {
      /\.xml$/ => $path,
      default  => "${path}/${name}.xml",
    }

    # generate and manage context configuration
    concat { "${name} tomcat context":
      path   => $path_real,
      owner  => $::tomcat::tomcat_user_real,
      group  => $::tomcat::tomcat_group_real,
      mode   => '0600',
      order  => 'numeric',
      notify => Service[$::tomcat::service_name_real]
    }
```

**Comment by antoineco** [comment]
Hey Jason, thanks for the feedback. I'm fine with the modifications you suggest, could you just post a Hiera config example and show me how you generate your instances so that I can test things on my side? (and understand your setup a bit better)

Cheers

**Comment by hanej** [comment]
Cool.  Yeah that's no problem.  Let me finish up the documentation and I'll post it.

**Comment by antoineco** [comment]
The main implementation is done and released in v1.1.0. Improvement requests can be done by submitting new issues. Closing.
